src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/EditorStateModel.cs

```diff
@@ -0,0 +1,289 @@
+            return _windowFrame.CloseFrame((uint)__FRAMECLOSE.FRAMECLOSE_PromptSave) == VSConstants.S_OK;
+        }
+
+        /// <summary>
+        /// Starts the process of closing the project file editor, if one is open currently.
+        /// </summary>
+        public async Task DisposeEditorAsync()
+        {
+            lock (_lock)
+            {
+                if (_currentState == EditorState.NoEditor) return;
+
+                // Checking for potential dirty state and asking if the user wants to save their changes will have already occurred at this point.
+                // Just go to EditorClosing.
+                _currentState = EditorState.EditorClosing;
+            }
+
+            _projectFileModelWatcher?.Dispose();
+            _textBufferStateListener?.Dispose();
+            if (_frameEventsListener != null)
+            {
+                await _frameEventsListener.DisposeAsync().ConfigureAwait(false);
+            }
+            _textBufferManager?.Dispose();
+
+            _projectFileModelWatcher = null;
+            _frameEventsListener = null;
+            _textBufferStateListener = null;
+            _textBufferManager = null;
+
+            lock (_lock)
+            {
+                _currentState = EditorState.NoEditor;
+            }
+            return;
+        }
+
+        // Update Project File Logic
+
+        /// <summary>
+        /// Schedules the project file to be updated. This is called by the ProjectFileModelWatcher, which is within a Project write lock.
+        /// Because of the time-sensitive nature of that lock, instead of updating the file immediately we schedule an update as soon
+        /// as the JTF can run our code.
+        /// </summary>
+        public JoinableTask ScheduleProjectFileUpdate()
+        {
+            lock (_lock)
+            {
+                // If the current state is writing project file, we don't want to update now, as the project will not have been fully
+                // reloaded yet. We'll get called back again after the ProjectReloadManager is finished reloading the project.
+                // Additionally, if there is no editor open, or if the editor is closing, don't schedule an update
+                if (_currentState == EditorState.WritingProjectFile || _currentState == EditorState.NoEditor || _currentState == EditorState.EditorClosing) return null;
+                _currentState = EditorState.BufferUpdateScheduled;
+                return _threadingService.JoinableTaskFactory.RunAsync(UpdateProjectFileAsync);
+            }
+        }
+
+        /// <summary>
+        /// Updates the content of the project file to be the latest msbuild text.
+        /// </summary>
+        /// <returns></returns>
+        private async Task UpdateProjectFileAsync()
+        {
+            // We do this on the UI thread to ensure that between setting the state to a variant of ProjectFileChanging and setting the window
```

Comment:
You mean if the buffer has unsaved changes? https://github.com/dotnet/roslyn-project-system/pull/1027/files/97b7dc865a59d48bde63ee250478a8523cc8a8fe#diff-adc55e7e05276851946a52ccbf9295d7R101